rfcs/framework-0063-notification-system.md

```diff
@@ -0,0 +1,422 @@
+                    if not m.user_id:
+                        continue
+                    if self.roles and m.role not in self.roles:
+                        continue
+                    user = m.relations.user.dereference()
+                    notif_pref = user.preferences["notifications"]
+                    com_pref = notif_pref.get(comm.id)
+                    if com_pref:
+                        # check if notification is enabled/disabled for the member
+                        # if ...:
+                        #     continue
+                    ret.append(user)
+            else:
+                ret.append(rec)
+        return ret
+
+# invenio_users_resources/notifications.py
+class UserRecipient:
+    def __call__(self, notification, recipients: list)
+        ret = []
+        for rec in recipients:
+            if isinstance(rec, User):
+                if rec.preferences["notifications"]["enabled"]
+                    ret.append(rec.dump())
+            else:
+                ret.append(rec)
+        return ret
+
+#
+# Backends
+#
+# invenio_users_resources/notifications.py
+class UserEmailBackend:
+    def __call__(self, notification, recipients: list)
+        for rec in recipients:
+            user = rec.data
+            rec.backends.append({
+                "backend": "email",
+                "to": f"{user.profile.full_name} <{user.email}>",
+            })
+
+#
+# Usage in a service method
+#
+uow.register(NotificationOp(CommunityRecordInclusionNotificationBuilder.build(
+    {
+        "request": request,
+        "community": community,
+        "record": record,
+    }
+)))
+```
+
+#### Notification examples
+
+| Action             | Who triggers   | Who receives                                                                            |
+| ------------------ | -------------- | --------------------------------------------------------------------------------------- |
+| Comment create     | Creating user  | All involved in conversation                                                            |
+| Comment edit       | Editing user   | All involved in conversation                                                            |
+| Comment delete     | Deleting user  | All involved in conversation                                                            |
+|                    |                |                                                                                         |
+| Submission create  | Creating user  | Community owners/managers                                                               |
+| Submission accept  | Accepting user | User who created submission                                                             |
+| Submission decline | Declining user | User who created submission                                                             |
+| Submission edit    | Editing user   | Either user who created submission or Community owners/managers (depends on who edited) |
```

Comment:
I don't think we're able to detect this event. We should log an event on the request timeline when a user edits an associated record (i.e. when in submitted state). But we don't want 20 events logged when a user edit the record and press save.